lib/Conversion/PolynomialToStandard/PolynomialToStandard.cpp

```diff
@@ -802,6 +803,263 @@ func::FuncOp PolynomialToStandard::buildPolynomialModFunc(FunctionType funcType,
+    if (!polyTy) {
+      op.emitError(
+          "Can't directly lower for a tensor of polynomials. "
+          "It is expected that it should be lowered to be done elementwise.");
+      return failure();
+    }
+
+    RingAttr ring = polyTy.getRing();
+    auto inputType = convertPolynomialType(polyTy);
+    // Retrieve the coefficients of the polynomial
+    auto inputVec = b.create<ToTensorOp>(inputType, adaptor.getOperands()[0]);
+
+    // Cast to intermediate type to avoid integer overflow during arithmetic
+    auto intermediateType =
+        inputType.clone(inputType.getShape(), polymulIntermediateType(polyTy));
+    Value initialValue =
+        b.create<arith::ExtUIOp>(intermediateType, inputVec.getResult());
+
+    // Create cmod for modulo arithmetic ops
+    auto cmod = ring.coefficientModulus();
+    Value cMod = b.create<arith::ConstantOp>(
+        intermediateType.getElementType(),
+        IntegerAttr::get(
+            intermediateType.getElementType(),
+            cmod.sextOrTrunc(intermediateType.getElementTypeBitWidth())));
+
+    // Compute the number of stages required to compute the NTT
+    auto degree = intermediateType.getShape()[0];
+    unsigned stages = (unsigned)std::log2((double)degree);
+
+    // Precompute the roots
+    auto root = ring.primitive2NthRoot().value();
+    auto rootsType = intermediateType.clone({degree});
+    auto roots = b.create<arith::ConstantOp>(
+        rootsType,
+        DenseElementsAttr::get(rootsType, precomputeRoots(root, cmod, degree)));
+
+    // Here is a slightly modified implementation of the standard iterative NTT
+    // computation using Cooley-Turkey butterfly. For reader reference:
+    // https://doi.org/10.1007/978-3-031-46077-7_22, and,
+    // https://doi.org/10.1109/ACCESS.2023.3294446
+    //
+    // We modify the standard implementation by pre-computing the root
+    // exponential values during compilation instead of doing so at runtime.
+    //
+    // Let roots be a tensor of <n x ix> where roots[i] = \psi^i, n be the
+    // degree of the polynomial and a be the coefficients. Then we implement the
+    // following:
+    //
+    // def fastNTT(a, n, cmod, roots):
+    //  a = BitReverseOrder(a)
+    //  m = n / 2
+    //  for (s = 0; s < log2(n); s++):
+    //    for (k = 0; k < n / m; k++):
+    //      for (j = 0; j < m / 2; j++):
+    //        A = a[k + j]
+    //        B = roots[(2 * j + 1) * n / m] * a[k + j + m / 2]
+    //        a[k + j] = A + B mod cmod
+    //        a[k + j + m / 2] = A - B mod cmod
+    //      end
+    //    end
+    //    m = m * 2
+    //  end
+
+    // Convert the input into Reverse-Bit Order
+    initialValue = computeReverseBitOrder(b, intermediateType, initialValue);
+
+    // Initialize the variables
+    Value initialBatchSize = b.create<arith::ConstantOp>(
+        IndexType::get(b.getContext()), b.getIndexAttr(2));
+
+    Value zero = b.create<arith::ConstantOp>(IndexType::get(b.getContext()),
```

Comment:
You can make these declarations slightly cleaner with the index-specialized helper

```
builder.create<arith::ConstantIndexOp>(loc, X)
```

Where `X` is the desired index value.

Reply:
There are similar things elsewhere for ConstantIntOp